### PR TITLE
[Feature] Add CFG parallel to Omnigen2 

### PR DESCRIPTION
**Pull request by zzhuoxin1508**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
- Add CFG parallel support to OmniGen2 pipeline (`cfg_parallel_size=2` for text-to-image, `cfg_parallel_size=3` for image editing with ref image)
- Fall back to sequential CFG when  image guidance is enabled without sufficient ranks (`cfg_world_size < 3`)
## Test Plan
- text2image
python text_to_image.py \
    --model "OmniGen2/OmniGen2" \
    --prompt "A classroom with bright lighting and wooden desks." \
    --negative-prompt "(((deformed))), blurry, over saturation, bad anatomy, disfigured, poorly drawn face, mutation, mutated, (extra_limb), (ugly), (poorly drawn hands), fused fingers, messy drawing, broken legs censor, censored, censor_bar" \
    --num-inference-steps 50 \
    --seed 0 \
    --guidance-scale 5.0 \
    --cfg-parallel-size 2 \
    --output /workspace/outputs/image_t2icfg3.png

- image2image
-python image_edit.py \
    --image /workspace/image1.png \
    --model "OmniGen2/OmniGen2" \
    --prompt "Change the background to classroom." \
    --negative-prompt "(((deformed))), blurry, over saturation, bad anatomy, disfigured, poorly drawn face, mutation, mutated, (extra_limb), (ugly), (poorly drawn hands), fused fingers, messy drawing, broken legs censor, censored, censor_bar" \
    --num-inference-steps 50 \
    --seed 0 \
    --guidance-scale 5.0 \
    --guidance-scale-2 2.0 \
    --cfg-parallel-size 3 \
    --output /workspace/outputs/image_edit.png
### Text-to-Image (1024×1024,)

| cfg_parallel_size | GPUs | Generation Time | Speedup |
|---|---|---|---|
| 1 | 1 | 54.4s | 1.0x |
| 2 | 2 | 27.3s | **2.0x** |

Output images are visually identical across sequential and parallel modes.

### Image Editing (inputsize=1696×2528)

| cfg_parallel_size | GPUs | Generation Time | Speedup |
|---|---|---|---|
| 1 | 1 | 90.2s | 1.0x |
| 3 | 3 | 36.6s | **2.46x** |

Output images are visually identical across sequential and parallel modes.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by zzhuoxin1508** [comment]
@nussejzz

**Comment by zzhuoxin1508** [comment]
text2image output 
image_t2icfg3
image2image output
image_editcfg3

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py [hunk 1]
```diff
@@ -1170,72 +1175,183 @@ def processing(
...
+        ref_latents,
+        negative_prompt_embeds,
+        negative_prompt_attention_mask,
+        timesteps,
+        dtype,
+        step_func=None,
+    ) -> torch.Tensor:
+        """CFG parallel denoising loop: each rank computes one CFG branch, returns latents.
+
+        Rank 0: cond branch  (prompt_embeds,          ref_latents)
```
Comment (nussejzz) [inline vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py]: Delete the spaces

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py [hunk 2]
```diff
@@ -1170,72 +1175,183 @@ def processing(
...
+        Rank 1: uncond branch (negative_prompt_embeds, None)
+        Rank 2: ref branch    (negative_prompt_embeds, ref_latents)  -- only when image_guidance > 1
+        """
+        cfg_group = get_cfg_group()
+        cfg_rank = get_classifier_free_guidance_rank()
+        use_cfg_img = self.image_guidance_scale > 1.0
```
Comment (nussejzz) [inline vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1288]: You have already written:"use_cfg_img = self.image_guidance_scale > 1.0" in line 1179

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py [hunk 3]
```diff
@@ -1170,72 +1175,183 @@ def processing(
         )
         self._num_timesteps = len(timesteps)
 
-        for i, t in enumerate(timesteps):
-            model_pred = self.predict(
-                t=t,
+        cfg_world_size = get_classifier_free_guidance_world_size()
+        use_cfg_img = self.image_guidance_scale > 1.0
```
Comment (nussejzz) [inline vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1179]: I recommend add `use_cfg_text = self.text_guidance_scale > 1.0` as well, so that you can replace all the self.text_guidance_scale > 1.0 and self.image_guidance_scale > 1.0 below, which will enhance code readability

**Review by nussejzz — CHANGES_REQUESTED** [review]
PTAL @zzhuoxin1508

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py [hunk 4]
```diff
@@ -1170,72 +1175,183 @@ def processing(
...
                 )
-
-                model_pred = (
-                    model_pred_uncond
-                    + image_guidance_scale * (model_pred_ref - model_pred_uncond)
-                    + text_guidance_scale * (model_pred - model_pred_ref)
-                )
-            elif text_guidance_scale > 1.0:
-                model_pred_uncond = self.predict(
+                gathered = cfg_group.all_gather(local_pred, separate_tensors=True)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1320]: `all_gather_into_tensor` requires contiguous input. `local_pred` coming out of `self.predict()` may not be contiguous depending on the transformer output layout. Add `local_pred = local_pred.contiguous()` before the all-gather, same way you already do for `latents` at line 1291.

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py [hunk 5]
```diff
@@ -1170,72 +1175,183 @@ def processing(
...
+            branch_attention_mask = negative_prompt_attention_mask
+            branch_ref_latents = None
+        else:
+            branch_prompt_embeds = negative_prompt_embeds
+            branch_attention_mask = negative_prompt_attention_mask
+            branch_ref_latents = ref_latents
+
+        for i, t in enumerate(timesteps):
+            in_cfg_range = self.cfg_range[0] <= i / len(timesteps) <= self.cfg_range[1]
+            use_cfg_this_step = in_cfg_range and self.text_guidance_scale > 1.0
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py]: `self.text_guidance_scale > 1.0` is always true inside `_processing_parallel` (the caller checks it in `cfg_parallel_ready`). This makes `use_cfg_this_step` equivalent to just `in_cfg_range`. Not a bug, but it's confusing — consider simplifying to `use_cfg_this_step = in_cfg_range`.

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py [hunk 6]
```diff
@@ -1170,72 +1175,183 @@ def processing(
...
-                model_pred = (
-                    model_pred_uncond
-                    + image_guidance_scale * (model_pred_ref - model_pred_uncond)
-                    + text_guidance_scale * (model_pred - model_pred_ref)
-                )
-            elif text_guidance_scale > 1.0:
-                model_pred_uncond = self.predict(
+                gathered = cfg_group.all_gather(local_pred, separate_tensors=True)
+                model_pred, model_pred_uncond = gathered[0], gathered[1]
+                if use_cfg_img_this_step and len(gathered) > 2:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py]: Nit: `len(gathered) > 2` is always true when `use_cfg_img_this_step` is true, since the caller guarantees `cfg_world_size >= 3` whenever `use_cfg_img`. The double-check reads like there's a case where it could be false. Drop it or add a comment explaining it's a defensive check.

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py [hunk 7]
```diff
@@ -1170,72 +1175,183 @@ def processing(
...
+                    model_pred = (
+                        model_pred_uncond
+                        + self.image_guidance_scale * (model_pred_ref - model_pred_uncond)
+                        + self.text_guidance_scale * (model_pred - model_pred_ref)
+                    )
+                else:
+                    model_pred = model_pred_uncond + self.text_guidance_scale * (model_pred - model_pred_uncond)
+                latents = self.scheduler.step(model_pred, t, latents, return_dict=False)[0]
+            else:
+                # Outside CFG interval: all ranks use cond branch, no comm
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1333]: Outside the CFG range every rank computes the same cond prediction independently — wasted FLOPs on ranks 1+. Consider having only rank 0 run `predict` and broadcasting `model_pred`, like you already do for the initial latents sync.

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py [hunk 8]
```diff
@@ -1170,72 +1175,183 @@ def processing(
...
+        #   rank 2+: ref branch if image guidance enabled, else uncond branch
+        if cfg_rank == 0:
+            branch_prompt_embeds = prompt_embeds
+            branch_attention_mask = prompt_attention_mask
+            branch_ref_latents = ref_latents
+        elif cfg_rank == 1:
+            branch_prompt_embeds = negative_prompt_embeds
+            branch_attention_mask = negative_prompt_attention_mask
+            branch_ref_latents = None
+        else:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1301]: When `cfg_world_size == 3` but `use_cfg_img` is False (T2I with `--cfg-parallel-size 3`), rank 2 runs the ref branch `(negative_prompt_embeds, ref_latents)` every step, and the result in `gathered[2]` is silently discarded. Either warn/error when `cfg_parallel_size > 2` for T2I, or have rank 2 mirror rank 1 to avoid wasted compute.

**Review by lishunyang12 — COMMENTED** [review]
Left a few comments on the parallel path.

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py (hunk 7 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1333]: Thanks,but data is the same on all cards, if we only run it on Rank 0, we’d just be adding an extra broadcast step. it wouldn't really save any time.

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py (hunk 8 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1301]: Since ref_latents is set to None when there's no input image, the data on Rank 1 and Rank 2 is exactly the same

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py (hunk 6 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py]: removed

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py (hunk 5 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py]: removed

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py (hunk 4 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1320]: done

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py (hunk 1 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py]: done

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py (hunk 3 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1179]: line1209-1206，I’m using the processed function after the check, not the raw function that was passed. So, I think we can leave it

File: vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py (hunk 2 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/pipeline_omnigen2.py:1288]: Intentionally kept it here to avoid adding an extra parameter to _processing_parallel.

**Comment by zzhuoxin1508** [comment]
@princepride Could you please review this

**Review by princepride — CHANGES_REQUESTED** [review]
I noticed that this model also need set cfg_p=3, right now we have a pr about refactor the cfg_p: https://github.com/vllm-project/vllm-omni/pull/2063, can you cooperate with him and try to use our own implementation of cfg_p.

**Comment by zzhuoxin1508** [comment]
> I noticed that this model also need set cfg_p=3, right now we have a pr about refactor the cfg_p: #2063, can you cooperate with him and try to use our own implementation of cfg_p.

Thanks, I'll look into it

**Comment by TKONIY** [comment]
> > I noticed that this model also need set cfg_p=3, right now we have a pr about refactor the cfg_p: #2063, can you cooperate with him and try to use our own implementation of cfg_p.
> 
> Thanks, I'll look into it

There are similar problems in integrating existing cfg_p (even after #2063) with Omnigen2 and DreamID-Omni: cfg_p now only supports defining no more than 2 branches, while Omnigen2 got 3 branches and DreamID-Omni got 4 branches.

Further refractor is needed to support multi-branches (>2) diffusion models. It will be good for us to have a discussion about a more general cfg parallel api, and the plan to refractor DreamID-Omni and Omnigen2 on it.

@zzhuoxin1508 
@princepride

**Comment by wtomin** [comment]
A recent PR changed the diffusion features docs strucure. Pls PTAL #1928.

**Comment by wtomin** [comment]
Missing e2e test for CFG parallelism. Please add a test that covers `--cfg-parallel-size=2` (Excluding  `--cfg-parallel-size=3` for now, because our nightly CI runs on two-gpus devices . For L4 test, please refer to #1832 . 

 Documentation incomplete:
1. Feature support table not updated for OmniGen2 CFG parallel
2. Usage example missing for --cfg-parallel-size=2/3 in `examples/offline_inference/image_to_image/image_to_image.md` and `examples/online_serving/image_to_image/image_to_image.md`

Can you also report the peak VRAM usage in your PR body?

**Comment by zzhuoxin1508** [comment]
> Missing e2e test for CFG parallelism. Please add a test that covers `--cfg-parallel-size=2` (Excluding `--cfg-parallel-size=3` for now, because our nightly CI runs on two-gpus devices . For L4 test, please refer to #1832 .
> 
> Documentation incomplete:
> 
> 1. Feature support table not updated for OmniGen2 CFG parallel
> 2. Usage example missing for --cfg-parallel-size=2/3 in `examples/offline_inference/image_to_image/image_to_image.md` and `examples/online_serving/image_to_image/image_to_image.md`
> 
> Can you also report the peak VRAM usage in your PR body?

Thanks for the review!
I'm currently working on a CFG parallel refactor in PR #2423 (N-branch CFG dispatch). I'll rebase this PR on top of that once it's merged, then add the  test, update the documentation and feature support table, and report peak VRAM usage in the PR  @wtomin

**Comment by zzhuoxin1508** [comment]
done in #2423
